### ISSUE TITLE
External Wifi Adapter Support

### ISSUE THREAD

**Issue by morpheuslord** [microsoft/WSL#8240]
hello, 
I am a hacker and I find using WSL over any other virtualization easy and when I am doing WIFI pen testing I need to use a external adapter which can support monitor mode but I looked up every single blog on the internet and none have pleasing results please help.
Please respond As Soon As Possible

Thank you
Morpheuslord

**Comment by akuropka** [comment]
https://github.com/dorssel/usbipd-win for attaching external USB devices. You should’ve come across if you checked every single piece of the i-net.

**Comment by morpheuslord** [comment]
Well I have tried it and this also has a lot of issues like I could not understand how should I update my kernel and secondly the build libraries for the adapter also won't install as the general libraries for the specific kernel is not present in the repository and in short I wanted to say that:

> kernel update or upgrade is a hassle 
> kernel repository not found
> mirrors not getting installed

The above won't allow me to link the usb adapter and them install the drivers for the adapters.

Thank you and hope you understand 

> 

>

**Comment by EDLLT** [comment]
> Well I have tried it and this also has a lot of issues like I could not understand how should I update my kernel and secondly the build libraries for the adapter also won't install as the general libraries for the specific kernel is not present in the repository and in short I wanted to say that:
> 
> > kernel update or upgrade is a hassle
> > kernel repository not found
> > mirrors not getting installed
> 
> The above won't allow me to link the usb adapter and them install the drivers for the adapters.
> 
> Thank you and hope you understand
> 
> > 
> 
> >

Hey bro, I am facing the exact same problem
did you find any solution to the problem?

Currently, I am messing around with the WSL kernel build trying to get my driver load in but with no avail

**Comment by morpheuslord** [comment]
No bro

**Comment by EDLLT** [comment]
> No bro

I MADE IT! IT WORKED!!!

THIS WAS ONE OF THE MOST COMPLICATED THINGS IN MY LIFE PROBABLY LOL

Okay so, holy crap that wasn't easy I had to stay up late till night for about 4 days trying to figure out how to do this as I had never compiled a kernel before

Now, everything works wlan0 is appearing as an external wifi adapter, injections work etc etc

I am gonna create a fork of that original project put everything there along with how to do it

ngl, I thought that it was impossible and was just playing around, and suddenly everything clicked and it worked

**Comment by EDLLT** [comment]
Here
https://github.com/EDLLT/WSL2-Linux-Kernel

**Comment by morpheuslord** [comment]
bro can you make a detailed a youtube vid on the installation and everything this will be a life save for hackers man

**Comment by GamerGirlandCo** [comment]
> bro can you make a detailed a youtube vid on the installation and everything this will be a life save for hackers man

just clone https://github.com/microsoft/WSL2-Linux-Kernel.git and apply [this patch](https://github.com/microsoft/WSL2-Linux-Kernel/commit/76598c510caee04aee422301182bdf0c8320cd42.patch) :)

**Comment by morpheuslord** [comment]
Gonna try it and will update of it worked

**Comment by morpheuslord** [comment]
> 

I am a complete noob when it comes to working with makefiles and their patches 😢 help please

**Comment by morpheuslord** [comment]
@akuropka it's not like I don't understand the process or how it's done but when it comes to C programming and concepts like bug fixing in C is really not the strong suite I tried compiling and made sure it works for my WLS once but I could not successfully imply these patches I always ran into some weird errors which had no certain answer.

**Comment by ishaan-kapoor** [comment]
hey, could you get it to work?

**Comment by morpheuslord** [comment]
nope not working.

**Comment by TheAjaykrishnanR** [comment]
It works ! for anybody wondering. Tried it with rtl8821au. You are a legend @EDLLT

**Comment by TheAjaykrishnanR** [comment]
> nope not working.

https://youtu.be/l7VQQJdePlk

**Comment by morpheuslord** [comment]
> > nope not working.
> 
> https://youtu.be/l7VQQJdePlk

hey so I tried it but was not able to find the drivers for Qualcomm Atheros AR9271 to run it according to the video 🥹

**Comment by TheAjaykrishnanR** [comment]
> > > nope not working.
> > 
> > 
> > https://youtu.be/l7VQQJdePlk
> 
> hey so I tried it but was not able to find the drivers for Qualcomm Atheros AR9271 to run it according to the video 🥹

https://cateee.net/lkddb/web-lkddb/ATH9K_HTC.html - i think its in there by default, https://github.com/EDLLT/WSL2-Linux-Kernel/tree/linux-msft-wsl-5.10.y/drivers/net/wireless/ath/ath9k

![image](https://github.com/microsoft/WSL/assets/110657810/3d2fbe04-6b0c-49ca-85ac-5d572a66faf4)

can you try enabling this ?

**Comment by morpheuslord** [comment]
> > > > nope not working.
> > > 
> > > 
> > > https://youtu.be/l7VQQJdePlk
> > 
> > 
> > hey so I tried it but was not able to find the drivers for Qualcomm Atheros AR9271 to run it according to the video 🥹
> 
> https://cateee.net/lkddb/web-lkddb/ATH9K_HTC.html - i think its in there by default, https://github.com/EDLLT/WSL2-Linux-Kernel/tree/linux-msft-wsl-5.10.y/drivers/net/wireless/ath/ath9k
> 
> ![image](https://private-user-images.githubusercontent.com/110657810/339840575-3d2fbe04-6b0c-49ca-85ac-5d572a66faf4.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTgzODQwMDUsIm5iZiI6MTcxODM4MzcwNSwicGF0aCI6Ii8xMTA2NTc4MTAvMzM5ODQwNTc1LTNkMmZiZTA0LTZiMGMtNDljYS04NWFjLTVkNTcyYTY2ZmFmNC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNjE0JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDYxNFQxNjQ4MjVaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1lMzNjOWUwMjg1M2E0N2IzMDljNzFiMmQ2NzI2ZWQ4OGI5YWZiZDIzZDk4M2EzOTM3ZDE5YTNlMTY3NTVhY2Q3JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.lkdL5KYu9ixm5R-IhodbmfKzgu7Csn1xZcJqikMDk40)
> 
> can you try enabling this ?

I had to elaborate, i did this but this is the issue during compilation "
  AR      init/built-in.a
  LD      vmlinux.o
ld: warning: arch/x86/lib/retpoline.o: missing .note.GNU-stack section implies executable stack
ld: NOTE: This behaviour is deprecated and will be removed in a future version of the linker
  MODPOST vmlinux.symvers
  MODINFO modules.builtin.modinfo
  GEN     modules.builtin
  LD      .tmp_vmlinux.btf
ld: warning: arch/x86/lib/retpoline.o: missing .note.GNU-stack section implies executable stack
ld: NOTE: This behaviour is deprecated and will be removed in a future version of the linker
ld: warning: .tmp_vmlinux.btf has a LOAD segment with RWX permissions
  BTF     .btf.vmlinux.bin.o
  LD      .tmp_vmlinux.kallsyms1
ld: warning: .btf.vmlinux.bin.o: missing .note.GNU-stack section implies executable stack
ld: NOTE: This behaviour is deprecated and will be removed in a future version of the linker
ld: warning: .tmp_vmlinux.kallsyms1 has a LOAD segment with RWX permissions
  KSYMS   .tmp_vmlinux.kallsyms1.S
  AS      .tmp_vmlinux.kallsyms1.S
  LD      .tmp_vmlinux.kallsyms2
ld: warning: .btf.vmlinux.bin.o: missing .note.GNU-stack section implies executable stack
ld: NOTE: This behaviour is deprecated and will be removed in a future version of the linker
ld: warning: .tmp_vmlinux.kallsyms2 has a LOAD segment with RWX permissions
  KSYMS   .tmp_vmlinux.kallsyms2.S
  AS      .tmp_vmlinux.kallsyms2.S
  LD      vmlinux
ld: warning: .btf.vmlinux.bin.o: missing .note.GNU-stack section implies executable stack
ld: NOTE: This behaviour is deprecated and will be removed in a future version of the linker
ld: warning: vmlinux has a LOAD segment with RWX permissions
  BTFIDS  vmlinux
FAILED: load BTF from vmlinux: Unknown error -22make: *** [Makefile:1179: vmlinux] Error 255

"

**Comment by TheAjaykrishnanR** [comment]
@morpheuslord ignoring the compiler's warning flags, this is the main error : 
`FAILED: load BTF from vmlinux: Unknown error -22make: *** [Makefile:1179: vmlinux] Error 255`
Where are you running this on ? Do you have around 10 GBs of RAM ?

**Comment by morpheuslord** [comment]
> @morpheuslord ignoring the compiler's warning flags, this is the main error : `FAILED: load BTF from vmlinux: Unknown error -22make: *** [Makefile:1179: vmlinux] Error 255` Where are you running this on ? Do you have around 10 GBs of RAM ?

on a VM and I have 16Gb of RAM on it.

**Comment by TheAjaykrishnanR** [comment]
> > @morpheuslord ignoring the compiler's warning flags, this is the main error : `FAILED: load BTF from vmlinux: Unknown error -22make: *** [Makefile:1179: vmlinux] Error 255` Where are you running this on ? Do you have around 10 GBs of RAM ?
> 
> on a VM and I have 16Gb of RAM on it.

can you try changing `CONFIG_DEBUG_INFO_BTF=y` to `CONFIG_DEBUG_INFO_BTF=n` in Microsoft/config-wsl ? Are you really giving 16 gbs of ram on your VM ?

**Comment by morpheuslord** [comment]
> > > @morpheuslord ignoring the compiler's warning flags, this is the main error : `FAILED: load BTF from vmlinux: Unknown error -22make: *** [Makefile:1179: vmlinux] Error 255` Where are you running this on ? Do you have around 10 GBs of RAM ?
> > 
> > on a VM and I have 16Gb of RAM on it.
> 
> can you try changing `CONFIG_DEBUG_INFO_BTF=y` to `CONFIG_DEBUG_INFO_BTF=n` in Microsoft/config-wsl ? Are you really giving 16 gbs of ram on your VM ? 

Yes I have 16GB on a VM, its currently late in the night so I will test it and update it tomorrow.

**Comment by TheAjaykrishnanR** [comment]
> > > > @morpheuslord ignoring the compiler's warning flags, this is the main error : `FAILED: load BTF from vmlinux: Unknown error -22make: *** [Makefile:1179: vmlinux] Error 255` Where are you running this on ? Do you have around 10 GBs of RAM ?
> > > 
> > > 
> > > on a VM and I have 16Gb of RAM on it.
> > 
> > 
> > can you try changing `CONFIG_DEBUG_INFO_BTF=y` to `CONFIG_DEBUG_INFO_BTF=n` in Microsoft/config-wsl ? Are you really giving 16 gbs of ram on your VM ?
> 
> Yes I have 16GB on a VM, its currently late in the night so I will test it and update it tomorrow.

sure ! im building it myself, will share it if it works

**Comment by morpheuslord** [comment]
@TheAjaykrishnanR  yup it's built with no errors, but wsl won't boot. It has now been 30min and there has been no reaction from wsl.

**Comment by TheAjaykrishnanR** [comment]
> @TheAjaykrishnanR yup it's built with no errors, but wsl won't boot. It has now been 30min and there has been no reaction from wsl.

can you run `wsl -l -v` to make sure you are running wsl 2 ?
anyways i had built it without changing that flag, check this kernel image after extracting the archive : https://www.dropbox.com/scl/fi/3cm7twaf4inlf4zqwviid/kernel_plus_modules_ath9k.tar.gz?rlkey=987wkudxvusrqms8ezzulfz5v&st=pge88leq&dl=0

**Comment by morpheuslord** [comment]
@

> **[TheAjaykrishnanR](/TheAjaykrishnanR) ** commented

It worked, thanks man. I dont know what was wrong with my current build but yours did work.

**Comment by TheAjaykrishnanR** [comment]
> @
> 
> > **[TheAjaykrishnanR](/TheAjaykrishnanR) ** commented
> 
> It worked, thanks man. I dont know what was wrong with my current build but yours did work. now its booting. but drivers wont work.
> 
> ```
> $ iwconfig
> lo        no wireless extensions.
> 
> bond0     no wireless extensions.
> 
> dummy0    no wireless extensions.
> 
> tunl0     no wireless extensions.
> 
> sit0      no wireless extensions.
> 
> eth0      no wireless extensions.
> ```
> 
> I lost hope at this point 🥹 .
> 
> dmesg | grep ath
> 
> ```
> [   15.420949] usb 1-1: ath9k_htc: Firmware ath9k_htc/htc_9271-1.4.0.fw requested
> [   15.421674] usbcore: registered new interface driver ath9k_htc
> [   15.422262] usb 1-1: Direct firmware load for ath9k_htc/htc_9271-1.4.0.fw failed with error -2
> [   15.422953] usb 1-1: ath9k_htc: Firmware htc_9271.fw requested
> [   15.424404] usb 1-1: ath9k_htc: Failed to get firmware htc_9271.fw
> [   15.424946] usb 1-1: ath9k_htc: USB layer deinitialized
> ```
> 
> I installed the firmware `firmware-atheros` and restarted wsl but still no luck.

the .ko driver has to be manually loaded each time you log into wsl using modprobe

**Comment by TheAjaykrishnanR** [comment]
> @
> 
> > **[TheAjaykrishnanR](/TheAjaykrishnanR) ** commented
> 
> It worked, thanks man. I dont know what was wrong with my current build but yours did work. now its booting. but drivers wont work.
> 
> ```
> $ iwconfig
> lo        no wireless extensions.
> 
> bond0     no wireless extensions.
> 
> dummy0    no wireless extensions.
> 
> tunl0     no wireless extensions.
> 
> sit0      no wireless extensions.
> 
> eth0      no wireless extensions.
> ```
> 
> I lost hope at this point 🥹 .
> 
> dmesg | grep ath
> 
> ```
> [   15.420949] usb 1-1: ath9k_htc: Firmware ath9k_htc/htc_9271-1.4.0.fw requested
> [   15.421674] usbcore: registered new interface driver ath9k_htc
> [   15.422262] usb 1-1: Direct firmware load for ath9k_htc/htc_9271-1.4.0.fw failed with error -2
> [   15.422953] usb 1-1: ath9k_htc: Firmware htc_9271.fw requested
> [   15.424404] usb 1-1: ath9k_htc: Failed to get firmware htc_9271.fw
> [   15.424946] usb 1-1: ath9k_htc: USB layer deinitialized
> ```
> 
> I installed the firmware `firmware-atheros` and restarted wsl but still no luck.

bro did you follow the entire tutorial ? Did you install the built modules ? copy the wsl2-modules into /lib. You have to use modprobe to load the .ko file each time you log onto wsl. Also use USB IPD to share the adapter with wsl. Follow the steps exactly

**Comment by morpheuslord** [comment]
@TheAjaykrishnanR  yes its loaded.

**Comment by TheAjaykrishnanR** [comment]
@morpheuslord do you have the `htc_9271.fw` file in `/lib/firmware` ?

**Comment by morpheuslord** [comment]
> @morpheuslord do you have the `htc_9271.fw` file in `/lib/firmware` ?
> 
> Anyways, I added the firmware directly to the kernel using `CONFIG_EXTRA_FIRMWARE` and compiled it again.
> 
> Use this kernel and it should hopefully work : https://www.dropbox.com/scl/fi/ik09bxk2uga3w3pniw6oe/kernel-plus-modules-ath9k-htc9271-fw.tar.gz?rlkey=uulfkbsonwe9hlwero73kv6ca&st=u1eltkxr&dl=0

Yes I tried both types of the .fx files but now worked I managed to get it to load but the dmesg says another story.

**Comment by morpheuslord** [comment]
> > > @morpheuslord do you have the `htc_9271.fw` file in `/lib/firmware` ?
> > > Anyways, I added the firmware directly to the kernel using `CONFIG_EXTRA_FIRMWARE` and compiled it again.
> > > Use this kernel and it should hopefully work : https://www.dropbox.com/scl/fi/ik09bxk2uga3w3pniw6oe/kernel-plus-modules-ath9k-htc9271-fw.tar.gz?rlkey=uulfkbsonwe9hlwero73kv6ca&st=u1eltkxr&dl=0
> > 
> > 
> > Yes I tried both types of the .fx files but now worked I managed to get it to load but the dmesg says another story.
> 
> you dont need the .fw files anymore, its embedded in the kernel image

your new kernel worked, as I was talking about before.
https://github.com/microsoft/WSL/issues/8240#issuecomment-2172324254

Thanks for that man, really appreciate it 👍.

**Comment by snehashis365** [comment]
> It works ! for anybody wondering. Tried it with rtl8821au. You are a legend @EDLLT

Hey man I have been trying to compile for this chipset but I keep running into memory bottleneck since you have already compiled for this can you share me the final tarball with the 8821au module, I tried by disabling the debug btf flag but it won't boot in that case

**Comment by TheAjaykrishnanR** [comment]
> > It works ! for anybody wondering. Tried it with rtl8821au. You are a legend @EDLLT
> 
> Hey man I have been trying to compile for this chipset but I keep running into memory bottleneck since you have already compiled for this can you share me the final tarball with the 8821au module, I tried by disabling the debug btf flag but it won't boot in that case

yeah here you go : https://www.dropbox.com/scl/fi/ex9q3ba9rqz3tt3w37k0l/wsl2-kernel-plus-modules-rtl8821au.tar.gz?rlkey=yhbjc00cls7iecu2u2b97h8ak&st=672dtq1d&dl=0

**Comment by snehashis365** [comment]
> 
> yeah here you go : https://www.dropbox.com/scl/fi/ex9q3ba9rqz3tt3w37k0l/wsl2-kernel-plus-modules-rtl8821au.tar.gz?rlkey=yhbjc00cls7iecu2u2b97h8ak&st=672dtq1d&dl=0

Thanks it worked!!

**Comment by Geofferey** [comment]
I was able to sucessfully compile the latest WSL kernel relatively easily. However, I experienced issue where flipping into monitor mode would crash WSL. Turns out the driver I had used wasn't good for whatever reason.

Here is the driver that worked for me:
https://github.com/ivanovborislav/rtl8821au

The driver I initially chose:
https://github.com/aircrack-ng/rtl8812au

**Comment by mieucun123** [comment]
> It works ! for anybody wondering. Tried it with rtl8821au. You are a legend @EDLLT

can you please help me with the TL-WN722N V1?

**Comment by TheAjaykrishnanR** [comment]
> > It works ! for anybody wondering. Tried it with rtl8821au. You are a legend @EDLLT
> 
> can you please help me with the TL-WN722N V1?

Isn't the chipset atheros 9271 ? or is it realtek ?

**Comment by mieucun123** [comment]
> > > It works ! for anybody wondering. Tried it with rtl8821au. You are a legend @EDLLT
> > 
> > 
> > can you please help me with the TL-WN722N V1?
> 
> Isn't the chipset atheros 9271 ? or is it realtek ?

It atheros 9271

**Comment by TheAjaykrishnanR** [comment]
@h471x if your ralink usb adapter is using rt2x00 then it should

**Comment by h471x** [comment]
> [@h471x](https://github.com/h471x) if your ralink usb adapter is using rt2x00 then it should

@TheAjaykrishnanR actually it is using ``mt7601u`` , any chance it should work ?

And by the way I wanna ask if the WSL2 kernel will eventually downgrade ( e.g. mine's from 5.15 but in the video I saw it has downgraded to 5.10 ) ??

**Comment by TheAjaykrishnanR** [comment]
@h471x 

https://github.com/morrownr/USB-WiFi/blob/main/home/USB_WiFi_Adapters_that_are_supported_with_Linux_in-kernel_drivers.md :

> chipset -  Mediatek mt7601u - N150 - USB 2 - supported in-kernel since Linux kernel 4.2 (2015)
> Note: the mt7601u driver only supports managed and monitor modes (no AP mode).

guess it is supported natively in the kernel

you can also read more about it hear : https://github.com/morrownr/USB-WiFi/discussions/95

about the kernel version, you could certainly build the newer kernel version 6 with the appropriate patches applied, though i did not find any particular reason to do so since you can always switch back to the newer one with wslconfig anyway once you are done with the pentesting

**Comment by h471x** [comment]
@TheAjaykrishnanR 

I guess I'll have to compile the custom Kernel,  
the ``mt7601u`` driver is indeed **supported** on ``Linux`` Kernel ( I use it all the time on a VM )  
but I guess that is NOT the case in WSL, I attached it on kali-linux WSL2 then I got those : 

<details>
  <summary>Commands & Outputs</summary>

  ```bash
  $ cmd.exe /c usbipd list
  Connected:
  BUSID  VID:PID    DEVICE                                                        STATE
  1-4    148f:7601  802.11n USB Wireless LAN Card                                 Attached
  1-6    13d3:56a8  USB2.0 HD UVC WebCam                                          Not shared
  1-7    0bda:0129  Realtek USB 2.0 Card Reader                                   Not shared
  1-10   8087:0aaa  Intel(R) Wireless Bluetooth(R)                                Not shared
  
  Persisted:
  GUID                                  DEVICE
  
  $ lsusb
  Bus 001 Device 001: ID 1d6b:0002 Linux Foundation 2.0 root hub
  Bus 001 Device 005: ID 148f:7601 Ralink Technology, Corp. MT7601U Wireless Adapter
  Bus 002 Device 001: ID 1d6b:0003 Linux Foundation 3.0 root hub

  $ iwconfig
  lo        no wireless extensions.
  
  eth0      no wireless extensions.
  
  $ dmesg | grep mt7601u
  
  $ sudo modprobe mt7601u
  modprobe: FATAL: Module mt7601u not found in directory /lib/modules/5.15.167.4-microsoft-standard-WSL2
  ```
</details>



> about the kernel version, you could certainly build the newer kernel version 6 with the appropriate patches applied, though i did not find any particular reason to do so since you can always switch back to the newer one with wslconfig anyway once you are done with the pentesting

About this you made it very clear to me, thank you !

**Comment by TheAjaykrishnanR** [comment]
@h471x 

>modprobe: FATAL: Module mt7601u not found in directory /lib/modules/5.15.167.4-microsoft-standard-WSL2

you have to enable them in menuconfig when you are building the kernel modules for the `.ko` file to be build. They will be in `kernel/drivers/net/wireless/mediatek/mt7601u` in the folder you output the make command

**Comment by h471x** [comment]
@TheAjaykrishnanR 

Ah I see, okay thank you, I will try it then I'll give feedback when I'm done 👍

**Comment by h471x** [comment]
I got this error when running WSL2 with the custom kernel

```bash
A request to send or receive data was disallowed because the socket is not connected and (when sending on a datagram socket using a sendto call) no address was supplied.
Error code: Wsl/Service/CreateInstance/CreateVm/WSAENOTCONN

[process exited with code 4294967295 (0xffffffff)]
You can now close this terminal with Ctrl+D, or press Enter to restart.
```

**Comment by TheAjaykrishnanR** [comment]
> I got this error when running WSL2 with the custom kernel
> 
> A request to send or receive data was disallowed because the socket is not connected and (when sending on a datagram socket using a sendto call) no address was supplied.
> Error code: Wsl/Service/CreateInstance/CreateVm/WSAENOTCONN
> 
> [process exited with code 4294967295 (0xffffffff)]
> You can now close this terminal with Ctrl+D, or press Enter to restart.

hmm seems to be an issue with later wsl versions and newer kernels : 

https://github.com/microsoft/WSL/issues/11840
https://github.com/microsoft/WSL/issues/11791

you could see if this works (add it to .wslconfig) : 

https://github.com/microsoft/WSL/issues/11791#issuecomment-2257225030
```
[wsl2]
kernelCommandLine=noxsave
```

**Comment by h471x** [comment]
It does work

I set .wslconfig like so : 

```bash
[wsl2]
kernelCommandLine=noxsave
```
Then I got :

```bash
C:\Users\Hatix Ntsoa>wsl
winux:/mnt/host/c/Users/Hatix Ntsoa#
```

**Comment by TheAjaykrishnanR** [comment]
@h471x the kernel should be the custom one `5.10.102.1-microsoft-standard-WSL2+` unless the one you have currently is the patched one and the kernel module should be in `/lib/modules/5.10.102.1-microsoft-standard-WSL2+/kernel/drivers/net/wireless/mediatek/mt7601`

**Comment by h471x** [comment]
Still got errors ! 😅
I can't load the custom kernel and I don't know why
